# Recitation 02：从表达式到完整程序

**依据 Lecture 02 与 Recitation 02；中文讲解下接英文翻译对照，保留英文原题。**

完成本讲义后，你应能：

- 先判断表达式的计算顺序，再同时说出结果和 Python 类型。
- 分清字符串字面量、变量和类型转换，解释正常输出或报错的原因。
- 跟踪多重赋值，区分 `print` 的内容、分隔符和结束符。
- 拆读格式说明符，并把百分比、温度换算和配方比例写成输入 → 处理 → 输出程序。

建议 **75–90 分钟**：概念复习 15 分钟；题 1–2 15 分钟；题 3–4 15 分钟；题 5–6 15 分钟；题 7–9 20 分钟；回顾 5 分钟。自学时按小节暂停，先预测、再运行、最后用一句话解释“为什么”。课堂时间不足时，题 6 的组合格式和应用题的额外测试可留作课后练习。

教师版包含逐项答案、易错点和已验证输出；学生版保留同一组原题、概念提示和可运行的 TODO。学生版初始 TODO 单元格故意不完成计算，可直接 **Restart Kernel and Run All**；填题后再运行，才能检验自己的答案。应用题使用可编辑的输入字符串，避免运行 notebook 时等待键盘；对应行旁边给出 `input()` 替换方式。

**来源版本与页码：** Lecture 02 是原生 PPTX，共 55 张 slide；引用中的 Slide N 是演示文稿顺序。Recitation 02 是 4 页 PDF。书籍为 Gaddis *Starting Out With Python*, Revel 6th ed.；`PDF p.` 一律指从 1 开始的物理页，`printed p.` 才是书内印刷页。逐题引用与哈希保存在同目录的 [source-map.md](source-map.md) / [source-map.json](source-map.json)。`direct` 表示直接讲到；`prerequisite` 表示只提供必要背景，具体组合由本讲义推导。

**原资料提示：** Lecture 02 Slide 55 写的是 “Read chapter 3”，但本次练习主要对应书籍 Chapter 2；这里保留阅读要求并标出不一致，不替老师更改作业。Lecture Slides 20、24、33、49 有个别演算或注释笔误；本讲义的演示以可运行 Python 为准，具体核对见教师说明。题 2 的后续章节引用只用于查 `str()` 和重复语法，不要求学习文件、列表或循环。

> **English translation:**
> # Recitation 02: From Expressions to Complete Programs
>
> Based on Lecture 02 and Recitation 02. Each Chinese explanation is followed by an English translation block, while the original exercise wording is preserved in English.
>
> After completing this handout, you should be able to:
>
> - determine the evaluation order of an expression and state both its result and its Python type;
> - distinguish string literals, variables, and type conversions, and explain either normal output or the reason for an error;
> - trace multiple assignment and distinguish the content, separator, and ending used by `print`;
> - read a format specifier piece by piece and turn percentage, temperature-conversion, and recipe-ratio problems into input → processing → output programs.
>
> Suggested time: **75–90 minutes**—15 minutes for concept review; 15 minutes for Exercises 1–2; 15 minutes for Exercises 3–4; 15 minutes for Exercises 5–6; 20 minutes for Exercises 7–9; and 5 minutes for the recap. When studying independently, pause after each subsection: predict first, run second, and finally explain *why* in one sentence. If class time is short, the combined formatting cases in Exercise 6 and the extra tests for the application problems can be assigned after class.
>
> The instructor version contains item-by-item answers, common mistakes, and verified output. The student version keeps the same original questions, concept prompts, and executable TODO cells. Its initial TODO cells intentionally leave calculations incomplete, so **Restart Kernel and Run All** is safe; rerun after filling them in to test your answers. Application exercises use editable input strings so the notebook does not wait for keyboard input; each relevant line shows how to replace the string with `input()`.
>
> **Source versions and page numbering.** Lecture 02 is a native 55-slide PPTX; “Slide N” means the presentation order. Recitation 02 is a four-page PDF. The book is Gaddis, *Starting Out With Python*, Revel 6th ed. `PDF p.` always means the one-based physical PDF page, while `printed p.` means the page number printed in the book. Exercise-level references and hashes are stored in [source-map.md](source-map.md) and [source-map.json](source-map.json). `direct` means the source directly teaches the point; `prerequisite` means it supplies necessary background and this handout derives the specific combination.
>
> **Note about the original materials.** Lecture 02 Slide 55 says “Read chapter 3,” although these exercises mainly align with Chapter 2. This handout preserves that instruction, records the mismatch, and does not silently change the assignment. Slides 20, 24, 33, and 49 contain isolated calculation or annotation errors; the executable Python demonstrations here are authoritative for this handout, and the instructor note records the details. The later-chapter references in Exercise 2 are used only to locate `str()` and repetition syntax; students are not expected to study files, lists, or loops here.
>
> **How to read common symbols aloud**
>
> | Symbol | English reading (IPA) | Reading in this notebook |
> |---|---|---|
> | `+` | **plus** /plʌs/ | “plus”; string `+` may be called the concatenation operator |
> | `-` | **minus** /ˈmaɪnəs/ | “minus”; before one value, “negative” or “unary minus” |
> | `*` | **asterisk** /ˈæstərɪsk/ | in arithmetic, “times” or “multiplied by” |
> | `/` | **slash** /slæʃ/ | in arithmetic, “divided by”; in Python, the true-division operator |
> | `//` | **double slash** /ˈdʌbəl slæʃ/ | “floor-divided by”; the floor-division operator |
> | `%` | **percent sign** /pərˈsent saɪn/ | for a percentage, “percent”; in Python arithmetic, “modulo” or “mod” |
> | `=` | **equals sign** /ˈiːkwəlz saɪn/ | in math, “equals”; in Python assignment, “is assigned” or “gets” |
> | `==` | **double equals** /ˈdʌbəl ˈiːkwəlz/ | the equality operator; read “is equal to” |
> | `(` `)` | **left/right parenthesis** /pəˈrenθəsɪs/ | together, “parentheses” |
> | `[` `]` | **left/right square bracket** /skwer ˈbrækɪt/ | square brackets |
> | `{` `}` | **left/right curly brace** /ˈkɜːrli breɪs/ | braces around an f-string replacement field |
> | `'` | **single quote** /ˈsɪŋɡəl kwoʊt/ | sometimes “apostrophe” when it is text content |
> | `"` | **double quote** /ˈdʌbəl kwoʊt/ | a string delimiter or a quote character |
> | `\` | **backslash** /ˈbækˌslæʃ/ | begins an escape sequence; `\n` is “backslash n,” `\t` is “backslash t” |
> | `,` | **comma** /ˈkɑːmə/ | separates arguments or assignment targets |
> | `:` | **colon** /ˈkoʊlən/ | introduces the format specifier in an f-string field |
> | `.` | **dot** /dɑːt/ | in a number, “decimal point”; in `.2f`, “dot two f” |
> | `_` | **underscore** /ˈʌndərskɔːr/ | joins words in an identifier |
> | `<` `>` | **less-than / greater-than sign** | in a format specifier, “left-align / right-align marker” |
> | `|` | **vertical bar** /ˈvɜːrtɪkəl bɑːr/ | a visible boundary or separator here; in a shell pipeline it is usually called a “pipe” /paɪp/ |
> | `!` | **exclamation mark** /ˌekskləˈmeɪʃən mɑːrk/ | punctuation or a literal output character |
> | `#` | **hash** /hæʃ/ | begins a Python comment; also called “number sign” in other contexts |
> | `$` | **dollar sign** /ˈdɑːlər saɪn/ | a currency symbol |
> | `°` | **degree sign** /dɪˈɡriː saɪn/ | read “degrees,” as in “25 degrees Celsius” |
> | `→` | **right arrow** /raɪt ˈæroʊ/ | read “goes to” or “becomes” |
> | `...` | **ellipsis** /ɪˈlɪpsɪs/ | omitted text or a placeholder, depending on context |


## 热身 A：先看运算符，再看类型

整数通常是 `int`，带小数点的数通常是 `float`。`/` 做真正的除法，即使两个整数能整除，结果也为 `float`。`//` 向负无穷取整；这与 `int()` 把小数部分朝 0 去掉不同。对这里的非零除数，余数满足 `a == (a // b) * b + (a % b)`；结果不为 0 时，余数与除数同号。

先算括号，再算幂，再算乘除、整除、取余，最后加减。同一优先级的乘除取余从左往右。不要只写数字而忘了类型：一项出现浮点数可能改变最终类型。下面是热身例子，数值与原题不同。

> **English translation:**
> ## Warm-up A: Read the Operators, Then the Types
>
> Integers are usually `int` values, and numbers containing a decimal point are usually `float` values. `/` (read “slash,” or “divided by” in arithmetic) performs true division; even when two integers divide evenly, the result is a `float`. `//` (read “double slash,” or “floor-divided by”) rounds the quotient toward negative infinity. This differs from `int()`, which removes the fractional part toward zero. For the nonzero divisors used here, the remainder satisfies `a == (a // b) * b + (a % b)`: read `==` as “is equal to,” `*` as “times,” `+` as “plus,” and `%` as “modulo.” When the remainder is nonzero, it has the same sign as the divisor.
>
> Evaluate parentheses first, then exponentiation, then multiplication, true division, floor division, and modulo, and finally addition and subtraction. Operators at the same multiplication/division level are evaluated from left to right. Do not report only the number and forget its type: the presence of a floating-point value may change the final type. The next cell is a warm-up example whose values differ from the original exercise.


In [1]:
print(29 / 6, type(29 / 6).__name__)
print(29 // 6, 29 % 6)
print(-29 // 6, int(-29 / 6))  # 向下取整 vs 朝 0 截断
print(4 + 6 * 2, (4 + 6) * 2)


4.833333333333333 float
4 5
-5 -4
16 20


## 热身 B：值、名字与类型

`"8"` 是字符串，`8` 是整数。字符串的 `+` 连接字符串；数值的 `+` 做加法。Python 不会替你猜测应该把哪一边转换。`int()`、`float()` 把合适的字符串变成数；`str()` 把值变成字符串。字符串重复需要整数次数。

变量名不用引号；有引号时得到的是那几个字符本身。`x = y` 把当时的 `y` 值赋给 `x`，不会建立以后自动同步的公式。多重赋值先求出整个右边的旧值，再按左边名字的顺序绑定。

> **English translation:**
> ## Warm-up B: Values, Names, and Types
>
> `"8"` is a string, whereas `8` is an integer. For strings, `+` (read “plus”) concatenates; for numbers, it adds. Python does not guess which operand you intended to convert. `int()` and `float()` convert suitable strings to numbers; `str()` converts a value to a string. String repetition requires an integer repetition count.
>
> A variable name is written without quotation marks. With quotes, the result is those literal characters. In `x = y`, read `=` as “is assigned” or “gets”: the current value of `y` is assigned to `x`; this does not create a formula that automatically stays synchronized. Multiple assignment first evaluates all old values on the right-hand side and then binds them to the names on the left in order.


In [2]:
text_count = "8"
print(text_count + "2", int(text_count) + 2)
print("la" * 3)
x = 9
y = x
x = 2
print(x, y)
x, y = y, x  # 先得到旧的右边，再赋给左边
print(x, y)


82 10
lalala
2 9
9 2


## 热身 C：字符串边界与输出边界

字符串开头的引号必须由同类引号结束；另一种引号可以作为内容。`\'` 和 `\"` 表示引号字符，`\n` 是换行，`\t` 是制表符。制表符移动到下一个制表位，显示距离受当前位置和环境影响，**不是固定几个空格**。

`print(value1, value2, sep=..., end=...)` 先把值变成文本，再在相邻参数之间放 `sep`，最后加 `end`。默认 `sep=' '`、`end='\n'`；字符串里面原有的空格不会被删除。`repr()` 是观察工具：它用带引号、带转义的形式描述文本，这些外层引号不是原始 `print` 自动打印的内容。

> **English translation:**
> ## Warm-up C: String Boundaries and Output Boundaries
>
> The quote that starts a string must be closed by the same kind of quote; the other kind may appear as content. `\'` is read “backslash single quote,” `\"` is “backslash double quote,” `\n` is “backslash n” (a newline), and `\t` is “backslash t” (a tab). A tab advances to the next tab stop, so its displayed width depends on the current position and environment; it is **not a fixed number of spaces**.
>
> `print(value1, value2, sep=..., end=...)` first converts values to text, places `sep` between adjacent arguments, and appends `end`. Read the commas as “comma,” `=` as assignment in these keyword arguments, and `...` as “ellipsis.” The defaults are `sep=' '` and `end='\n'`; spaces already inside a string are not removed. `repr()` is an inspection tool: it describes text with surrounding quotes and escape sequences. Those outer quotes are not characters automatically printed by the original `print` call.


In [3]:
print("red", "blue", sep=" / ", end="!\n")
message = "X\tY\nZ"
print(message)
print("辅助 repr：", repr(message))


red / blue!
X	Y
Z
辅助 repr： 'X\tY\nZ'


## 热身 D：格式化是生成显示文本

f-string 的 `{表达式:格式说明}` 先计算值，再生成字符串。字段宽度是**最小**宽度，内容太长通常不会被剪掉。`<` 左对齐，`>` 右对齐；紧挨对齐符之前的字符可以作为填充。

精度取决于被格式化的类型：浮点数的 `.2f` 表示小数点后两位；浮点数只写 `.2`、省略 `f` 时，使用类似通用格式的有效数字规则，不等于固定两位小数。字符串的 `.2` 最多保留两个字符。先读值的类型，再解释精度；不要把宽度当精度。格式化不改变原变量的数值。

> **English translation:**
> ## Warm-up D: Formatting Produces Display Text
>
> In an f-string, `{expression:format_specifier}` first evaluates the expression and then produces a string. Read `{` and `}` as “left/right curly brace” and `:` as “colon.” Field width is a **minimum** width; content that is too long is normally not cut off. `<` is the left-alignment marker and `>` is the right-alignment marker. A character immediately before an alignment marker can be used as the fill character.
>
> Precision depends on the formatted value's type. For a float, `.2f` (read “dot two f”) means two digits after the decimal point. If a float uses `.2` without `f`, Python applies significant-digit rules similar to general formatting; it does not mean exactly two decimal places. For a string, `.2` keeps at most two characters. Read the value's type before interpreting precision, and do not confuse width with precision. Formatting does not change the original variable's numeric value.


In [4]:
amount = 6.789
print("|" + f"{amount:9.2f}" + "|")
print("|" + f"{'owl':.>7}" + "|")
print("|" + f"{'sparrow':<7.2}" + "|")
print(amount)  # 原来的值没有变


|     6.79|
|....owl|
|sp     |
6.789


## 热身 E：输入 → 处理 → 输出

`input()` 总是返回字符串。先按意义选 `int()`（整数数量）或 `float()`（允许小数），再计算，最后打印结果和单位。先用能心算的输入检查公式方向和比例。

> **English translation:**
> ## Warm-up E: Input → Processing → Output
>
> `input()` always returns a string. Choose `int()` for an integer quantity or `float()` when decimals are allowed, then calculate, and finally print the result with its unit. Start with an input you can calculate mentally so you can check the direction and scale of the formula. Read `→` as “goes to.”


In [5]:
distance_text = "3.5"  # 交互时替换为 input("距离（km）：")
distance_km = float(distance_text)
distance_m = distance_km * 1000
print(f"距离：{distance_m:.1f} m")


距离：3500.0 m


**教师辅助工具（无需学生记忆）。** 下方工具仅把实际打印捕获为 `repr`，帮助检查换行、制表符、前后空格与 `end`。练习答案中的原始语句仍会正常执行。这里的 `import`、函数、上下文管理器与 `exec` 是验证设施，不是本次要求掌握的语法；学生版会移除工具及其调用。报错演示使用 `try/except` 或 `compile`，让 notebook 继续执行；捕获后显示的 `error — ...` 是讲义的提示，不是原题语句产生的正常输出。

> **English translation:**
> **Instructor utility; students do not need to memorize it.** The utility below captures the actual printed text and displays its `repr`, making newlines, tabs, leading/trailing spaces, and `end` easier to inspect. The original statements in the exercise solutions still execute normally. The `import`, function definition, context manager, and `exec` call are verification infrastructure, not syntax students must learn in this recitation; the student version removes both the utility and its calls. Error demonstrations use `try/except` (read the slash as “or” here) or `compile` so the notebook can continue. A captured message such as `error — ...` is a handout annotation, not normal output from the original statement.


In [6]:
import io
from contextlib import redirect_stdout

def show_invisible(source, values=None):
    """仅供教师：执行同一段原题，显示实际标准输出的 repr。"""
    capture = io.StringIO()
    with redirect_stdout(capture):
        exec(source, {} if values is None else values.copy())
    print("\n[辅助 repr；外层引号不是原输出]", repr(capture.getvalue()))


## 1. 数值表达式：结果、类型和运算顺序

**原题（保留来源提取文本）**

```text
Exercises 1 – Numeric values and operations 
Evaluate the following expressions and specify their type.  
5 * -6 
 
34 / 10 
 
(34 // 10) * (1.0 + 2) 
 
34 % 10 
 
23 % 4.0 
 
25 + 30 / 6 
 
-7 % 3 
 
100 - 25 * 3 % 4
```

**Knowledge references / 知识点出处**

- Lecture: [*Lecture 02*](../../lecture/Lecture%2002.pptx), Slide 14 (PPTX), “Mixing numeric types” — **direct**；覆盖：numeric result types。
- Lecture: [*Lecture 02*](../../lecture/Lecture%2002.pptx), Slide 19 (PPTX), “Operators on int and float” — **direct**；覆盖：true and floor division / remainder operation。
- Lecture: [*Lecture 02*](../../lecture/Lecture%2002.pptx), Slide 20 (PPTX), “Operators on int and float” — **direct**；覆盖：negative-operand remainder。 This slide includes a negative-dividend modulo example; its separate 10 // 0.25 output contains a typo.
- Lecture: [*Lecture 02*](../../lecture/Lecture%2002.pptx), Slide 22 (PPTX), “Operators Precedence” — **direct**；覆盖：operator precedence。
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 2, §2.7: Performing Calculations, printed p. 90 (PDF p. 91) — **direct**；覆盖：numeric result types。 Read the general mixed-type rules together with the explicit / exception on PDF p.84.
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 2, §2.7: Performing Calculations, printed p. 83 (PDF p. 84) — **direct**；覆盖：true and floor division。
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 2, §2.7: Performing Calculations, printed p. 87 (PDF p. 88) — **direct**；覆盖：remainder operation。
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 2, §2.7: Performing Calculations, printed p. 83 (PDF p. 84) — **prerequisite**；覆盖：negative-operand remainder。 Teaches negative floor division, a prerequisite for interpreting negative modulo; does not explicitly demonstrate negative-dividend %.
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 2, §2.7: Performing Calculations, printed p. 84 (PDF p. 85) — **direct**；覆盖：operator precedence。

原题来源：[recitation/Recitaion 02.pdf](../../recitation/Recitaion%2002.pdf#page=1)，PDF p. 1（物理页）。

### 本题考察什么

这题同时检查“怎么算”和“算出来是什么类型”。8 个表达式依次涉及：① 负数参与乘法；② 真除法的返回类型；③ 括号、整除与混合类型；④ 整数余数；⑤ 浮点除数下的余数类型；⑥ 除法优先于加法；⑦ 负数的整除—余数关系；⑧ 乘法、取余、减法的先后。

**怎么做：** 每项先在纸上写出结果和类型，再运行；解释至少一步中间计算。

**来源与表述说明：** 负数取余的教材引用是 prerequisite：教材讲负数整除，本讲义再用商和余数的关系推导。教材 PDF p. 91 的一般类型规则要与 PDF p. 84 的 `/` 例外一起读。

> **English translation:**
> ## 1. Numeric Expressions: Results, Types, and Evaluation Order
>
> **Original exercise:** the English source text above is preserved verbatim.
>
> **What this exercise tests.** It checks both *how* an expression is evaluated and *what Python type* its result has. The eight expressions cover, in order: (1) multiplication involving a negative number; (2) the return type of true division; (3) parentheses, floor division, and mixed numeric types; (4) an integer remainder; (5) the remainder type with a floating-point divisor; (6) division before addition; (7) the floor-division/remainder relationship for a negative value; and (8) the ordering of multiplication, modulo, and subtraction.
>
> **How to proceed.** For each item, first write the result and type on paper, then run the cell. Explain at least one intermediate calculation.
>
> **Source and wording note.** The textbook reference for negative modulo is marked `prerequisite`: the book teaches negative floor division, and this handout derives the remainder using the quotient–remainder relationship. The general type rule on textbook PDF p. 91 must be read together with the `/` exception on PDF p. 84. Read `/` as “slash” or “divided by,” `//` as “double slash” or “floor-divided by,” `%` as “modulo,” and `*` as “times.”


### 1.1 `5 * -6`

预测：结果 = ___；类型 = ___。

> **English translation:**
> Prediction: result = ___; type = ___. Read `=` as “equals” in this answer blank.


In [7]:
value = 5 * -6
print(value, type(value).__name__)


-30 int


**逐项讲解：** 两个操作数都是整数；先理解 `-6` 是负数，再相乘，得到 `-30`，类型为 `int`。易错：不要把负号误当成乘法之后才执行的另一条语句。

> **English translation:**
> **Item explanation.** Both operands are integers. Interpret `-6` as negative six, then multiply to obtain `-30`, whose type is `int`. Common mistake: treating the minus sign as though it were a separate statement executed after multiplication. In arithmetic, read `*` as “times” and unary `-` as “negative.”


### 1.2 `34 / 10`

预测：结果 = ___；类型 = ___。

> **English translation:**
> Prediction: result = ___; type = ___. Read `=` as “equals” in this answer blank.


In [8]:
value = 34 / 10
print(value, type(value).__name__)


3.4 float


**逐项讲解：** `/` 做真正的除法，结果 `3.4`，类型为 `float`。即便能整除，`/` 的结果仍是 `float`，不是“有小数才变 float”。

> **English translation:**
> **Item explanation.** `/` (read “slash” or “divided by”) performs true division, producing `3.4` with type `float`. Even when the division is exact, `/` still returns a `float`; it is not the presence of a visible fractional part that turns the result into a float.


### 1.3 `(34 // 10) * (1.0 + 2)`

预测：结果 = ___；类型 = ___。

> **English translation:**
> Prediction: result = ___; type = ___. Read `=` as “equals” in this answer blank.


In [9]:
value = (34 // 10) * (1.0 + 2)
print(value, type(value).__name__)


9.0 float


**逐项讲解：** 左括号 `34 // 10` 得到整数 `3`；右括号 `1.0 + 2` 得到浮点数 `3.0`；最终 `9.0`，类型为 `float`。易错：只看到整除就把整个表达式判为 int。

> **English translation:**
> **Item explanation.** The left parentheses, `34 // 10`, produce the integer `3`; the right parentheses, `1.0 + 2`, produce the float `3.0`. The final result is `9.0`, with type `float`. Common mistake: seeing floor division and therefore classifying the whole expression as an `int`. Read `//` as “double slash” or “floor-divided by,” `+` as “plus,” and `*` as “times.”


### 1.4 `34 % 10`

预测：结果 = ___；类型 = ___。

> **English translation:**
> Prediction: result = ___; type = ___. Read `=` as “equals” in this answer blank.


In [10]:
value = 34 % 10
print(value, type(value).__name__)


4 int


**逐项讲解：** `34 = 3 * 10 + 4`，余数 `4`，类型为 `int`。`%` 在这里不是“百分比符号”。

> **English translation:**
> **Item explanation.** `34 = 3 * 10 + 4`, so the remainder is `4` and its type is `int`. Here `%` is the Python modulo/remainder operator, read “modulo” or “mod,” not a percentage sign.


### 1.5 `23 % 4.0`

预测：结果 = ___；类型 = ___。

> **English translation:**
> Prediction: result = ___; type = ___. Read `=` as “equals” in this answer blank.


In [11]:
value = 23 % 4.0
print(value, type(value).__name__)


3.0 float


**逐项讲解：** `23 = 5 * 4.0 + 3.0`，结果 `3.0`，类型为 `float`。余数即使没有非零小数部分，也不会因此自动变回 int。

> **English translation:**
> **Item explanation.** `23 = 5 * 4.0 + 3.0`, so the result is `3.0` with type `float`. A remainder does not automatically change back to `int` merely because its displayed value has no nonzero fractional digits. Read `%` as “modulo,” `*` as “times,” and `+` as “plus.”


### 1.6 `25 + 30 / 6`

预测：结果 = ___；类型 = ___。

> **English translation:**
> Prediction: result = ___; type = ___. Read `=` as “equals” in this answer blank.


In [12]:
value = 25 + 30 / 6
print(value, type(value).__name__)


30.0 float


**逐项讲解：** 先算 `30 / 6 = 5.0`，再算 `25 + 5.0 = 30.0`，类型为 `float`。不能把 `25 + 30` 先加起来。

> **English translation:**
> **Item explanation.** First evaluate `30 / 6 = 5.0`, then `25 + 5.0 = 30.0`; the result's type is `float`. Do not add `25 + 30` first. Read `/` as “divided by” and `+` as “plus.”


### 1.7 `-7 % 3`

预测：结果 = ___；类型 = ___。

> **English translation:**
> Prediction: result = ___; type = ___. Read `=` as “equals” in this answer blank.


In [13]:
value = -7 % 3
print(value, type(value).__name__)


2 int


**逐项讲解：** 先求向下取整的商：`-7 // 3 = -3`，不是朝 0 截断的 `-2`。再由 `-7 = (-3) * 3 + 2` 得到余数 `2`，类型为 `int`。这不是把正数余数直接加上负号。

> **English translation:**
> **Item explanation.** First find the floor-divided quotient: `-7 // 3 = -3`, not the toward-zero truncation `-2`. Then `-7 = (-3) * 3 + 2`, so the remainder is `2` with type `int`. This is not obtained by simply attaching a minus sign to the positive remainder. Read `//` as “floor-divided by” and `%` as “modulo.”


### 1.8 `100 - 25 * 3 % 4`

预测：结果 = ___；类型 = ___。

> **English translation:**
> Prediction: result = ___; type = ___. Read `=` as “equals” in this answer blank.


In [14]:
value = 100 - 25 * 3 % 4
print(value, type(value).__name__)


97 int


**逐项讲解：** 乘法与取余同级，先从左到右算 `25 * 3 = 75`，再算 `75 % 4 = 3`，最后 `100 - 3 = 97`，类型为 `int`。易错：先做减法，或者把 `%` 当成更低优先级。

> **English translation:**
> **Item explanation.** Multiplication and modulo have the same precedence, so evaluate left to right: `25 * 3 = 75`, then `75 % 4 = 3`, and finally `100 - 3 = 97`, with type `int`. Common mistakes are subtracting first or treating `%` as lower precedence. Read `*` as “times,” `%` as “modulo,” and `-` as “minus.”


## 2. 看输出：同一个符号，不同的类型

**原题（保留来源提取文本）**

```text
Exercise 2 - Looking for Prints  
Write out the output of each line in the space provided. If the line results in an error, 
write "error". 
print(5 + "5")  
 
 
    print(5 + int("5")) 
print(str(5) + int("5"))    
    print(15 % 2) 
print("hey" + "yo")  
 
    print(2 * "go") 
print("yo" * 4.0) 
 
 
    print("hi" + there) 
print("go" * 2 + "!" * 2)
```

**Knowledge references / 知识点出处**

- Lecture: [*Lecture 02*](../../lecture/Lecture%2002.pptx), Slide 15 (PPTX), “Data Type Conversion” — **direct**；覆盖：int conversion / str conversion。
- Lecture: [*Lecture 02*](../../lecture/Lecture%2002.pptx), Slide 19 (PPTX), “Operators on int and float” — **direct**；覆盖：numeric remainder。
- Lecture: [*Lecture 02*](../../lecture/Lecture%2002.pptx), Slide 25 (PPTX), “Operations Summary” — **direct**；覆盖：string concatenation and operand types。
- Lecture: [*Lecture 02*](../../lecture/Lecture%2002.pptx), Slide 24 (PPTX), “Operator on String” — **direct**；覆盖：string repetition / integer repetition count。 The slide directly shows both operand orders and a float-count TypeError; its separate Hello + ICP example has an incorrect displayed result.
- Lecture: [*Lecture 02*](../../lecture/Lecture%2002.pptx), Slide 33 (PPTX), “Variables” — **direct**；覆盖：unassigned variable names。
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 2, §2.6: Reading Input from the Keyboard, printed p. 78 (PDF p. 79) — **direct**；覆盖：int conversion。
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 6, §6.1: Introduction to File Input and Output, printed p. 362 (PDF p. 363) — **direct**；覆盖：str conversion。 Targeted later-chapter reference for str() only; file handling is outside this recitation.
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 2, §2.7: Performing Calculations, printed p. 87 (PDF p. 88) — **direct**；覆盖：numeric remainder。
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 2, §2.8: String Concatenation, printed p. 93 (PDF p. 94) — **direct**；覆盖：string concatenation and operand types。
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 8, §8.3: Testing, Searching, and Manipulating Strings, printed p. 500 (PDF p. 501) — **direct**；覆盖：string repetition。 Targeted later-chapter reference for string repetition only; loops and password validation are outside this recitation.
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 7, §7.2: Introduction to Lists, printed p. 406 (PDF p. 407) — **prerequisite**；覆盖：integer repetition count。 States the integer operand requirement for sequence repetition using lists; string repetition is separately demonstrated on PDF p.501.
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 2, §2.5: Variables, printed p. 68 (PDF p. 69) — **direct**；覆盖：unassigned variable names。

原题来源：[recitation/Recitaion 02.pdf](../../recitation/Recitaion%2002.pdf#page=1)，PDF p. 1（物理页）。

### 本题考察什么

这题检查你能否在运行前判断操作数类型。按原表格逐行、从左到右共有 9 项：① 数值和字符串能否相加；② `int` 转换后的加法；③ `str` 和 `int` 混用；④ 数值取余；⑤ 字符串连接；⑥ 整数重复次数；⑦ 看起来像整数的浮点次数；⑧ 带引号文本与未赋值名字；⑨ 重复与连接的组合顺序。

**怎么做：** 写正常输出或 error，并说明是哪一类类型/名字问题。把 9 行视为独立表达式；`there` 未预先定义。

**来源与表述说明：** 右列行首空格来自 PDF 表格排版，不是 Python 缩进。后续教材章节只定位个别语法；列表重复的引用是 integer repetition count 的 prerequisite。

> **English translation:**
> ## 2. Reading Output: The Same Symbol with Different Types
>
> **Original exercise:** the English source text above is preserved verbatim.
>
> **What this exercise tests.** It checks whether you can determine operand types before running code. Reading the original table row by row from left to right gives nine items: (1) whether a number and string can be added; (2) addition after `int` conversion; (3) mixing `str` and `int`; (4) numeric modulo; (5) string concatenation; (6) an integer repetition count; (7) a floating-point count that looks mathematically integral; (8) quoted text versus an unassigned name; and (9) the combined order of repetition and concatenation.
>
> **How to proceed.** Write the normal output or `error`, then identify the type or name problem. Treat all nine lines as independent expressions; `there` has not been defined in advance.
>
> **Source and wording note.** Leading spaces in the right column come from PDF table layout and are not Python indentation. Later textbook chapters are cited only to locate isolated syntax. The list-repetition source is a `prerequisite` for the integer repetition-count rule. Read `+` as “plus” or, for strings, “concatenation”; `*` as “times” or “repeated”; and `%` as “modulo.”


### 2.1

```python
print(5 + "5")
```

预测输出或 error，并写出判断依据。

> **English translation:**
> Predict the output or write `error`, and state the reason for your decision.


In [15]:
try:
    print(5 + "5")
except TypeError as error:
    print("error — TypeError:", error)


error — TypeError: unsupported operand type(s) for +: 'int' and 'str'


**逐项讲解：** `5` 是 int，`"5"` 是 str；两者不能直接用 `+`。结果为 error（TypeError），没有正常打印值。

> **English translation:**
> **Item explanation.** `5` is an `int`, while `"5"` is a `str`. They cannot be combined directly with `+` (read “plus”), so the result is an error (`TypeError`) and no value is printed normally.


### 2.2

```python
print(5 + int("5"))
```

预测输出或 error，并写出判断依据。

> **English translation:**
> Predict the output or write `error`, and state the reason for your decision.


In [16]:
print(5 + int("5"))


10


**逐项讲解：** `int("5")` 得到整数 5，数值相加后输出 `10`。转换的是值，不是把 `int` 这个词打印出来。

> **English translation:**
> **Item explanation.** `int("5")` produces the integer 5, and numeric addition prints `10`. The value is converted; the word `int` itself is not printed.


### 2.3

```python
print(str(5) + int("5"))
```

预测输出或 error，并写出判断依据。

> **English translation:**
> Predict the output or write `error`, and state the reason for your decision.


In [17]:
try:
    print(str(5) + int("5"))
except TypeError as error:
    print("error — TypeError:", error)


error — TypeError: can only concatenate str (not "int") to str


**逐项讲解：** 左边经过 `str` 后为字符串，右边经过 `int` 后为整数，仍是 str + int，得到 error（TypeError）。“两边都转换过”并不表示类型一致。

> **English translation:**
> **Item explanation.** The left operand becomes a string through `str`, while the right operand becomes an integer through `int`. The operation is still `str + int`, so it raises a `TypeError`. Converting both sides does not imply that the resulting types match. Read `+` as “plus.”


### 2.4

```python
print(15 % 2)
```

预测输出或 error，并写出判断依据。

> **English translation:**
> Predict the output or write `error`, and state the reason for your decision.


In [18]:
print(15 % 2)


1


**逐项讲解：** 数值取余，15 除以 2 余 1，因此输出 `1`。

> **English translation:**
> **Item explanation.** This is numeric remainder: 15 divided by 2 leaves 1, so the output is `1`. Read `%` as “modulo” or “mod.”


### 2.5

```python
print("hey" + "yo")
```

预测输出或 error，并写出判断依据。

> **English translation:**
> Predict the output or write `error`, and state the reason for your decision.


In [19]:
print("hey" + "yo")


heyyo


**逐项讲解：** 两个字符串直接连接，输出 `heyyo`，不会自动加空格；字符串分隔引号不属于输出。

> **English translation:**
> **Item explanation.** The two strings are concatenated directly, producing `heyyo`; Python does not insert a space automatically. The quotes that delimit each string are not part of the output. Read `+` here as the string-concatenation operator.


### 2.6

```python
print(2 * "go")
```

预测输出或 error，并写出判断依据。

> **English translation:**
> Predict the output or write `error`, and state the reason for your decision.


In [20]:
print(2 * "go")


gogo


**逐项讲解：** 整数在左侧也可以指定字符串重复次数，输出 `gogo`。

> **English translation:**
> **Item explanation.** An integer on the left can also specify a string repetition count, so the output is `gogo`. Read `*` as “repeated” in this string context.


### 2.7

```python
print("yo" * 4.0)
```

预测输出或 error，并写出判断依据。

> **English translation:**
> Predict the output or write `error`, and state the reason for your decision.


In [21]:
try:
    print("yo" * 4.0)
except TypeError as error:
    print("error — TypeError:", error)


error — TypeError: can't multiply sequence by non-int of type 'float'


**逐项讲解：** 虽然 4.0 的数学值是整数，它的 Python 类型仍是 float；字符串重复次数要求整数，得到 error（TypeError）。

> **English translation:**
> **Item explanation.** Although 4.0 is mathematically an integer value, its Python type is still `float`. A string repetition count must have an integer type, so the expression raises a `TypeError`. Read `*` as “repeated.”


### 2.8

```python
print("hi" + there)
```

预测输出或 error，并写出判断依据。

> **English translation:**
> Predict the output or write `error`, and state the reason for your decision.


In [22]:
try:
    print("hi" + there)
except NameError as error:
    print("error — NameError:", error)


error — NameError: name 'there' is not defined


**逐项讲解：** 先求变量 `there` 的值，但它未定义，所以得到 error（NameError）。不能擅自把 `there` 当成字符串字面量。

> **English translation:**
> **Item explanation.** Python first tries to retrieve the value of the variable `there`, but that name is undefined, so it raises a `NameError`. Do not silently reinterpret `there` as a string literal.


### 2.9

```python
print("go" * 2 + "!" * 2)
```

预测输出或 error，并写出判断依据。

> **English translation:**
> Predict the output or write `error`, and state the reason for your decision.


In [23]:
print("go" * 2 + "!" * 2)


gogo!!


**逐项讲解：** 先分别重复得到 `gogo` 和 `!!`，再连接，输出 `gogo!!`。没有额外空格。

> **English translation:**
> **Item explanation.** The two repetitions first produce `gogo` and `!!`; concatenation then produces `gogo!!`, with no extra spaces. Read `*` as “repeated,” `+` as “concatenated with,” and `!` as “exclamation mark.”


## 3. 引号与转义：代码文本和实际字符

**原题（保留来源提取文本）**

```text
Exercises 3 - Escape sequence 
Write the output, or Error if an error occurs. 
print("Welcome to \'ICP\'") 
 
print("Welcome to "ICP") 
 
print('Welcome to "ICP') 
print('Welcome to \"ICP') 
 
print("Welcome to \"ICP\"") 
 
print("Today is\t\"Tuesday\"\nWelcome to\t\"ICP\"")
```

**Knowledge references / 知识点出处**

- Lecture: [*Lecture 02*](../../lecture/Lecture%2002.pptx), Slide 9 (PPTX), “Strings” — **direct**；覆盖：string delimiters and syntax errors。
- Lecture: [*Lecture 02*](../../lecture/Lecture%2002.pptx), Slide 10 (PPTX), “Strings” — **direct**；覆盖：escaped quotation marks。
- Lecture: [*Lecture 02*](../../lecture/Lecture%2002.pptx), Slide 11 (PPTX), “Escape Sequence” — **direct**；覆盖：newline and tab characters。
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 2, §2.3: Displaying Output with the print Function, printed p. 63 (PDF p. 64) — **prerequisite**；覆盖：string delimiters and syntax errors。 Teaches matching delimiters and mixed quote styles; does not explicitly trace the malformed print line in this exercise.
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 2, §2.9: More About the print Function, printed p. 97 (PDF p. 98) — **direct**；覆盖：escaped quotation marks。
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 2, §2.9: More About the print Function, printed p. 96 (PDF p. 97) — **direct**；覆盖：newline and tab characters。

原题来源：[recitation/Recitaion 02.pdf](../../recitation/Recitaion%2002.pdf#page=1)，PDF p. 1（物理页）。

### 本题考察什么

6 行分别检查：① 双引号字符串内转义单引号；② 字符串定界符是否匹配；③ 单引号字符串内的普通双引号；④ 同一位置的转义双引号；⑤ 成对显示双引号；⑥ 引号、制表符、换行共同出现时的实际输出。判断错误时先检查语法，再讨论输出。

**怎么做：** 先数清字符串从哪里开始、在哪里结束；再把转义序列解释成实际字符。第 3、4 行虽然同处原表格一格，也分别完成。

> **English translation:**
> ## 3. Quotes and Escapes: Source-Code Text versus Actual Characters
>
> **Original exercise:** the English source text above is preserved verbatim.
>
> **What this exercise tests.** The six lines check, respectively: (1) an escaped single quote inside a double-quoted string; (2) whether string delimiters match; (3) an ordinary double quote inside a single-quoted string; (4) an escaped double quote in the same position; (5) a displayed pair of double quotes; and (6) the actual output when quotation marks, tabs, and a newline occur together. When deciding whether a line is erroneous, check syntax before discussing output.
>
> **How to proceed.** First identify exactly where the string begins and ends; then interpret every escape sequence as an actual character. Although source rows 3 and 4 appear in the same table cell, complete them separately. Read `\` as “backslash,” `\'` as “backslash single quote,” `\"` as “backslash double quote,” `\t` as “backslash t,” and `\n` as “backslash n.”


### 3.1

```python
print("Welcome to \'ICP\'")
```

In [24]:
print("Welcome to \'ICP\'")
# 教师观察辅助：显示同一 print 的实际输出，含末尾换行。
show_invisible('print("Welcome to \\\'ICP\\\'")')


Welcome to 'ICP'

[辅助 repr；外层引号不是原输出] "Welcome to 'ICP'\n"


**逐项讲解：** 双引号界定整个字符串，`\'` 产生一个单引号。实际输出 `Welcome to 'ICP'`；反斜杠不显示。此处单引号不需要转义也能作为内容，但原题写法有效。

> **English translation:**
> **Item explanation.** Double quotes delimit the whole string, and `\'` produces one single-quote character. The actual output is `Welcome to 'ICP'`; the backslash is not displayed. A single quote could appear here without escaping, but the source line as written is valid. Read `\` as “backslash” and `'` as “single quote.”


### 3.2

```python
print("Welcome to "ICP")
```

In [25]:
bad_code = 'print("Welcome to "ICP")'
try:
    compile(bad_code, "<exercise-3-line-2>", "exec")
except SyntaxError as error:
    print("error — SyntaxError:", error.msg)


error — SyntaxError: unterminated string literal (detected at line 1)


**逐项讲解：** 中间的双引号过早结束字符串，后面的 `ICP` 和末尾引号无法组成合法表达式，发生 SyntaxError。它不是运行后才产生的类型问题，原题没有正常输出。讲义用 compile 捕获错误，不改写原题。

> **English translation:**
> **Item explanation.** The middle double quote closes the string too early. The following `ICP` and final quote cannot form a valid expression, so Python raises a `SyntaxError`. This is not a type problem discovered after execution, and the original statement has no normal output. The handout catches the error with `compile` without rewriting the source line.


### 3.3

```python
print('Welcome to "ICP')
```

In [26]:
print('Welcome to "ICP')
# 教师观察辅助：显示同一 print 的实际输出，含末尾换行。
show_invisible('print(\'Welcome to "ICP\')')


Welcome to "ICP

[辅助 repr；外层引号不是原输出] 'Welcome to "ICP\n'


**逐项讲解：** 外层用单引号，里面的双引号是普通字符，因此输出 `Welcome to "ICP`，只有左边那一个双引号；不要自行补上原题没写的右双引号。

> **English translation:**
> **Item explanation.** Single quotes delimit the outer string, so the double quote inside it is an ordinary character. The output is `Welcome to "ICP`, containing only the left double quote. Do not add a right quote that does not appear in the source.


### 3.4

```python
print('Welcome to \"ICP')
```

In [27]:
print('Welcome to \"ICP')
# 教师观察辅助：显示同一 print 的实际输出，含末尾换行。
show_invisible('print(\'Welcome to \\"ICP\')')


Welcome to "ICP

[辅助 repr；外层引号不是原输出] 'Welcome to "ICP\n'


**逐项讲解：** `\"` 同样得到双引号字符，因此这一行与上一行输出完全相同。反斜杠改变源代码写法，不会作为字符留在输出中。

> **English translation:**
> **Item explanation.** `\"` also produces a double-quote character, so this line has exactly the same output as the previous one. The backslash changes how the source code is written; it does not remain as a character in the output. Read `\"` as “backslash double quote.”


### 3.5

```python
print("Welcome to \"ICP\"")
```

In [28]:
print("Welcome to \"ICP\"")
# 教师观察辅助：显示同一 print 的实际输出，含末尾换行。
show_invisible('print("Welcome to \\"ICP\\"")')


Welcome to "ICP"

[辅助 repr；外层引号不是原输出] 'Welcome to "ICP"\n'


**逐项讲解：** 两处 `\"` 各产生一个双引号，所以输出 `Welcome to "ICP"`。代码最外层的字符串引号不显示，内容里的两个双引号会显示。

> **English translation:**
> **Item explanation.** Each occurrence of `\"` produces one double-quote character, so the output is `Welcome to "ICP"`. The outer string delimiters are not displayed; the two double quotes inside the content are displayed.


### 3.6

```python
print("Today is\t\"Tuesday\"\nWelcome to\t\"ICP\"")
```

In [29]:
print("Today is\t\"Tuesday\"\nWelcome to\t\"ICP\"")
# 教师观察辅助：显示同一 print 的实际输出，含末尾换行。
show_invisible('print("Today is\\t\\"Tuesday\\"\\nWelcome to\\t\\"ICP\\"")')


Today is	"Tuesday"
Welcome to	"ICP"

[辅助 repr；外层引号不是原输出] 'Today is\t"Tuesday"\nWelcome to\t"ICP"\n'


**逐项讲解：** 第一行文本含 Tuesday，换行后第二行含 ICP；每行相应位置各有一个制表符。`\n` 是内容中的换行，print 的默认 end 还会在整个输出末尾补一个换行。repr 中的 `\t` 不能解释成固定数量的空格。

> **English translation:**
> **Item explanation.** The first output line contains “Tuesday,” and after a newline the second line contains “ICP”; each line has one tab at the corresponding position. `\n` (“backslash n”) is the newline inside the string, and the default `end` of `print` adds another newline only at the end of the entire output. A `\t` (“backslash t”) shown by `repr` must not be interpreted as a fixed number of spaces.


## 4. 追踪 Lady：多重赋值先读旧值

**原题（保留来源提取文本）**

```text
Exercise 4 | Variable assignment 
Consider the following variable assignments:  
 
Left = "Lady" 
 
Middle = "Jack" 
 
Right = "Ace"  
Figure out where the 'Lady' is (Left, Middle, Right) after running the following programs: 
Left, Right, Middle = Left, Middle, Right 
 
Left, Right, Middle = Middle, Right, Left 
 
Left, Middle, Right = Middle, Right, Left 
 
Left, Middle, Right = Middle, Right, Left 
Left, Middle, Right = Middle, Right, Left 
 
Left, Middle, Right = Middle, Right, Left 
Left, Middle, Right = Middle, Right, Left 
Left, Middle, Right = Middle, Right, Left 
 
Left, Middle, Right = Right, Right, Right 
Left, Middle, Right = Left, Left, Left 
Left, Middle, Right = Middle, Middle, Middle
```

**Knowledge references / 知识点出处**

- Lecture: [*Lecture 02*](../../lecture/Lecture%2002.pptx), Slide 33 (PPTX), “Variables” — **direct**；覆盖：multiple assignment。
- Lecture: [*Lecture 02*](../../lecture/Lecture%2002.pptx), Slide 29 (PPTX), “How do variables actually work?” — **prerequisite**；覆盖：old right-hand-side values before rebinding。 Teaches evaluate-then-bind for ordinary assignment; applying that order to every RHS value in a multiple assignment requires an explicit teaching step.
- Lecture: [*Lecture 02*](../../lecture/Lecture%2002.pptx), Slide 32 (PPTX), “Reassignment of Variables” — **direct**；覆盖：reassignment and loss of references。
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 2, §2.5: Variables, printed p. 69 (PDF p. 70) — **direct**；覆盖：multiple assignment。
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 2, §2.5: Variables, printed p. 69 (PDF p. 70) — **prerequisite**；覆盖：old right-hand-side values before rebinding。 Demonstrates multiple assignment of literals; does not explicitly explain taking all old RHS values before rebinding existing names.
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 2, §2.5: Variables, printed p. 71 (PDF p. 72) — **direct**；覆盖：reassignment and loss of references。

原题来源：[recitation/Recitaion 02.pdf](../../recitation/Recitaion%2002.pdf#page=2)，PDF p. 2（物理页）。

### 本题考察什么

这题考察多重赋值、左右位置的配对、逐行状态追踪，以及重新赋值后原值是否仍由某个变量保存。六段程序分别有 1、1、1、2、3、3 条赋值语句；不要把表格所有行连成一段程序。

**怎么做：** 每段都从 `Left="Lady"`、`Middle="Jack"`、`Right="Ace"` 开始。对每条语句先写右侧旧值，再按左侧出现顺序分配；最后回答 Lady 在哪里。如果没有变量再保存 Lady，可以回答“无”，无需勉强选三个位置之一。

**来源与表述说明：** 原题只列 Left/Middle/Right 选项；这里补充“无”作为必要的判定范围，不提前指定适用于哪段。关于先求全部 RHS 再绑定名字的引用包含 prerequisite，本讲义明确补充这一步推理。

> **English translation:**
> ## 4. Tracking Lady: Multiple Assignment Reads Old Values First
>
> **Original exercise:** the English source text above is preserved verbatim.
>
> **What this exercise tests.** It covers multiple assignment, positional pairing between the left and right sides, line-by-line state tracing, and whether an old value remains stored by any variable after reassignment. The six programs contain 1, 1, 1, 2, 3, and 3 assignment statements, respectively. Do not concatenate every row of the table into one program.
>
> **How to proceed.** Start each program with `Left="Lady"`, `Middle="Jack"`, and `Right="Ace"`. For each statement, first write down the old values on the right-hand side; then assign them according to the order of names on the left. Finally state Lady's location. If no variable stores Lady, answer “none”; do not force a choice among the three locations. In Python assignment, read `=` as “is assigned” or “gets,” and read commas as “comma.”
>
> **Source and wording note.** The original exercise lists only Left, Middle, and Right. This handout adds “none” as a necessary response category without revealing which program requires it. The source for evaluating the entire right-hand side before binding names is marked `prerequisite`; this handout makes that reasoning step explicit.


### 4.1 独立程序（1 条赋值）

```python
Left, Right, Middle = Left, Middle, Right
```

追踪列始终按 `Left | Middle | Right` 记录，最终 Lady 的位置：___。

> **English translation:**
> ### 4.1 Independent program (one assignment)
>
> Always record the trace in the fixed column order `Left | Middle | Right`; read each `|` as “vertical bar.” Final location of Lady: ___.


In [30]:
Left = "Lady"
Middle = "Jack"
Right = "Ace"
print("初始 (Left, Middle, Right):", Left, Middle, Right)
Left, Right, Middle = Left, Middle, Right
print("第 1 行后:", Left, Middle, Right)


初始 (Left, Middle, Right): Lady Jack Ace
第 1 行后: Lady Ace Jack


**逐项讲解：** 先取右侧旧值 `(Lady, Jack, Ace)`，但左侧顺序是 `(Left, Right, Middle)`，最终按固定列 `(Left, Middle, Right)` 看为 `(Lady, Ace, Jack)`；Lady 在 Left。易错：忽略左侧名字的顺序。

> **English translation:**
> **Item explanation.** First take the old right-hand-side values `(Lady, Jack, Ace)`. The left-hand order is `(Left, Right, Middle)`, so in the fixed display order `(Left, Middle, Right)` the final state is `(Lady, Ace, Jack)`; Lady is in Left. Common mistake: ignoring the order of names on the left.


### 4.2 独立程序（1 条赋值）

```python
Left, Right, Middle = Middle, Right, Left
```

追踪列始终按 `Left | Middle | Right` 记录，最终 Lady 的位置：___。

> **English translation:**
> ### 4.2 Independent program (one assignment)
>
> Always record the trace in the fixed column order `Left | Middle | Right`; read each `|` as “vertical bar.” Final location of Lady: ___.


In [31]:
Left = "Lady"
Middle = "Jack"
Right = "Ace"
print("初始 (Left, Middle, Right):", Left, Middle, Right)
Left, Right, Middle = Middle, Right, Left
print("第 1 行后:", Left, Middle, Right)


初始 (Left, Middle, Right): Lady Jack Ace
第 1 行后: Jack Lady Ace


**逐项讲解：** 先取右侧旧值 `(Jack, Ace, Lady)`，对应左侧 `(Left, Right, Middle)`，最终固定列为 `(Jack, Lady, Ace)`；Lady 在 Middle。易错：把第三个值误放进 Right。

> **English translation:**
> **Item explanation.** First take the old right-hand-side values `(Jack, Ace, Lady)` and match them to `(Left, Right, Middle)`. In the fixed column order the result is `(Jack, Lady, Ace)`; Lady is in Middle. Common mistake: putting the third value into Right.


### 4.3 独立程序（1 条赋值）

```python
Left, Middle, Right = Middle, Right, Left
```

追踪列始终按 `Left | Middle | Right` 记录，最终 Lady 的位置：___。

> **English translation:**
> ### 4.3 Independent program (one assignment)
>
> Always record the trace in the fixed column order `Left | Middle | Right`; read each `|` as “vertical bar.” Final location of Lady: ___.


In [32]:
Left = "Lady"
Middle = "Jack"
Right = "Ace"
print("初始 (Left, Middle, Right):", Left, Middle, Right)
Left, Middle, Right = Middle, Right, Left
print("第 1 行后:", Left, Middle, Right)


初始 (Left, Middle, Right): Lady Jack Ace
第 1 行后: Jack Ace Lady


**逐项讲解：** 右侧旧值为 `(Jack, Ace, Lady)`，左侧这次是标准顺序，最终 `(Jack, Ace, Lady)`；Lady 在 Right。不要用已更新的 Left 去计算右侧最后一项。

> **English translation:**
> **Item explanation.** The old right-hand-side values are `(Jack, Ace, Lady)`, and the left side is now in the standard order. The final state is `(Jack, Ace, Lady)`; Lady is in Right. Do not use an already updated value of Left when evaluating the last item on the right.


### 4.4 独立程序（2 条赋值）

```python
Left, Middle, Right = Middle, Right, Left
Left, Middle, Right = Middle, Right, Left
```

追踪列始终按 `Left | Middle | Right` 记录，最终 Lady 的位置：___。

> **English translation:**
> ### 4.4 Independent program (two assignments)
>
> Always record the trace in the fixed column order `Left | Middle | Right`; read each `|` as “vertical bar.” Final location of Lady: ___.


In [33]:
Left = "Lady"
Middle = "Jack"
Right = "Ace"
print("初始 (Left, Middle, Right):", Left, Middle, Right)
Left, Middle, Right = Middle, Right, Left
print("第 1 行后:", Left, Middle, Right)
Left, Middle, Right = Middle, Right, Left
print("第 2 行后:", Left, Middle, Right)


初始 (Left, Middle, Right): Lady Jack Ace
第 1 行后: Jack Ace Lady
第 2 行后: Ace Lady Jack


**逐项讲解：** 第 1 行后 `(Jack, Ace, Lady)`，第 2 行以这个新状态作为旧值，得到 `(Ace, Lady, Jack)`；Lady 在 Middle。每行内部同时绑定，行与行之间按顺序执行。

> **English translation:**
> **Item explanation.** After line 1 the state is `(Jack, Ace, Lady)`. Line 2 uses that new state as its old state and produces `(Ace, Lady, Jack)`; Lady is in Middle. Bindings occur simultaneously within each line, while separate lines execute in order.


### 4.5 独立程序（3 条赋值）

```python
Left, Middle, Right = Middle, Right, Left
Left, Middle, Right = Middle, Right, Left
Left, Middle, Right = Middle, Right, Left
```

追踪列始终按 `Left | Middle | Right` 记录，最终 Lady 的位置：___。

> **English translation:**
> ### 4.5 Independent program (three assignments)
>
> Always record the trace in the fixed column order `Left | Middle | Right`; read each `|` as “vertical bar.” Final location of Lady: ___.


In [34]:
Left = "Lady"
Middle = "Jack"
Right = "Ace"
print("初始 (Left, Middle, Right):", Left, Middle, Right)
Left, Middle, Right = Middle, Right, Left
print("第 1 行后:", Left, Middle, Right)
Left, Middle, Right = Middle, Right, Left
print("第 2 行后:", Left, Middle, Right)
Left, Middle, Right = Middle, Right, Left
print("第 3 行后:", Left, Middle, Right)


初始 (Left, Middle, Right): Lady Jack Ace
第 1 行后: Jack Ace Lady
第 2 行后: Ace Lady Jack
第 3 行后: Lady Jack Ace


**逐项讲解：** 三行后依次为 `(Jack, Ace, Lady)` → `(Ace, Lady, Jack)` → `(Lady, Jack, Ace)`；Lady 回到 Left。这是同一个循环移位做三次，而不是三次都读取最初的值。

> **English translation:**
> **Item explanation.** The states after the three lines are `(Jack, Ace, Lady)` → `(Ace, Lady, Jack)` → `(Lady, Jack, Ace)`; Lady returns to Left. This is the same cyclic shift performed three times, not three operations that all read the initial state. Read `→` as “goes to.”


### 4.6 独立程序（3 条赋值）

```python
Left, Middle, Right = Right, Right, Right
Left, Middle, Right = Left, Left, Left
Left, Middle, Right = Middle, Middle, Middle
```

追踪列始终按 `Left | Middle | Right` 记录，最终 Lady 的位置：___。

> **English translation:**
> ### 4.6 Independent program (three assignments)
>
> Always record the trace in the fixed column order `Left | Middle | Right`; read each `|` as “vertical bar.” Final location of Lady: ___.


In [35]:
Left = "Lady"
Middle = "Jack"
Right = "Ace"
print("初始 (Left, Middle, Right):", Left, Middle, Right)
Left, Middle, Right = Right, Right, Right
print("第 1 行后:", Left, Middle, Right)
Left, Middle, Right = Left, Left, Left
print("第 2 行后:", Left, Middle, Right)
Left, Middle, Right = Middle, Middle, Middle
print("第 3 行后:", Left, Middle, Right)


初始 (Left, Middle, Right): Lady Jack Ace
第 1 行后: Ace Ace Ace
第 2 行后: Ace Ace Ace
第 3 行后: Ace Ace Ace


**逐项讲解：** 第 1 行把当时 Right 的 Ace 同时赋给三个变量，因此状态成为 `(Ace, Ace, Ace)`；后两行仍然只能复制 Ace。Lady 已不在任何变量中，答案为“无”。变量数没有减少，但它们保存的值可能相同，原先的值不会自动保留下来。

> **English translation:**
> **Item explanation.** Line 1 assigns the current value of Right, Ace, to all three variables simultaneously, so the state becomes `(Ace, Ace, Ace)`. The next two lines can only copy Ace. Lady is no longer stored by any variable, so the answer is “none.” The number of variables has not decreased, but several variables may store the same value, and old values are not preserved automatically.


## 5. print 的 sep 与 end：逐字符拼接

**原题（保留来源提取文本）**

```text
Exercise 5 | Print arguments  
a = 5 
print("a", a, sep=' = ') 
 
a = 5 
print("a = ", a, sep="0000") 
 
a = 5 
print("a = ", a, sep="'", end="'") 
 
a, b = 'b', 'a' 
print(a, 'b', sep='>>>', end='\t') 
print('a', b, sep='***', end='\n\n') 
print(a, b, sep='\t')
```

**Knowledge references / 知识点出处**

- Lecture: [*Lecture 02*](../../lecture/Lecture%2002.pptx), Slide 44 (PPTX), “print function” — **direct**；覆盖：string literals versus variable values。
- Lecture: [*Lecture 02*](../../lecture/Lecture%2002.pptx), Slide 46 (PPTX), “print – sep and end arguments” — **direct**；覆盖：print separators and endings。
- Lecture: [*Lecture 02*](../../lecture/Lecture%2002.pptx), Slide 11 (PPTX), “Escape Sequence” — **direct**；覆盖：newline and tab characters。
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 2, §2.5: Variables, printed p. 71 (PDF p. 72) — **direct**；覆盖：string literals versus variable values。
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 2, §2.9: More About the print Function, printed p. 95 (PDF p. 96) — **direct**；覆盖：print separators and endings。
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 2, §2.9: More About the print Function, printed p. 96 (PDF p. 97) — **direct**；覆盖：newline and tab characters。

原题来源：[recitation/Recitaion 02.pdf](../../recitation/Recitaion%2002.pdf#page=3)，PDF p. 3（物理页）。

### 本题考察什么

这题把字符串/变量、分隔符和结束符放在一起。4 段程序依次检查：① 字面量与变量之间自定义分隔；② 原字符串尾空格和分隔字符串叠加；③ 分隔符与结束符中的引号是真实字符；④ 多次 print 因不同 end 连成哪些行。

**怎么做：** 原题以表格留白呈现，讲义将任务解读为“预测每段程序的完整输出”。每段独立执行，包含最后一段的全部 3 次 print；明确最后是否有换行。

**来源与表述说明：** 本题没有额外英文指令句；上面的中文任务说明是教学解释，不是增补伪造的原题文字。

> **English translation:**
> ## 5. `print` with `sep` and `end`: Constructing Output Character by Character
>
> **Original exercise:** the English source text above is preserved verbatim.
>
> **What this exercise tests.** It combines string literals versus variables, separators, and endings. The four programs check: (1) a custom separator between a literal and a variable; (2) the combination of a trailing space in the original string and a separator string; (3) quote characters used as the separator and ending; and (4) how several `print` calls form lines because they use different `end` values.
>
> **How to proceed.** The original is presented as a table with blank answer spaces, so the handout interprets the task as predicting each program's complete output. Execute each program independently, including all three `print` calls in the final program, and state whether a final newline is present.
>
> **Source and wording note.** The source includes no additional English instruction sentence; the Chinese task description above is pedagogical guidance, not fabricated source wording. Read `sep` aloud as “sep” or “separator,” `end` as “end,” `=` as “is assigned” in code, `\t` as “backslash t,” and `\n` as “backslash n.”


### 5.1

```python
a = 5
print("a", a, sep=' = ')
```

In [36]:
a = 5
print("a", a, sep=' = ')
show_invisible('a = 5\nprint("a", a, sep=\' = \')')


a = 5

[辅助 repr；外层引号不是原输出] 'a = 5\n'


**逐项讲解：** 第一个参数是字符 a，第二个参数是变量值 5；只在两者之间插入 ` = `，输出 `a = 5`，然后默认换行。

> **English translation:**
> **Item explanation.** The first argument is the character `a`, and the second argument is the variable value 5. The separator ` = ` is inserted only between them, so the output is `a = 5`, followed by the default newline. Read `=` in the displayed text as “equals.”


### 5.2

```python
a = 5
print("a = ", a, sep="0000")
```

In [37]:
a = 5
print("a = ", a, sep="0000")
show_invisible('a = 5\nprint("a = ", a, sep="0000")')


a = 00005

[辅助 repr；外层引号不是原输出] 'a = 00005\n'


**逐项讲解：** 第一个参数 `"a = "` 自带末尾空格，再插入四个 0，之后才是 5，完整文本为 `a = 00005`，然后换行。sep 不会替换或删除参数内的空格。

> **English translation:**
> **Item explanation.** The first argument, `"a = "`, already ends with a space. Python then inserts four zero characters before printing 5, so the complete text is `a = 00005`, followed by a newline. `sep` does not replace or delete spaces inside an argument.


### 5.3

```python
a = 5
print("a = ", a, sep="'", end="'")
```

In [38]:
a = 5
print("a = ", a, sep="'", end="'")
show_invisible('a = 5\nprint("a = ", a, sep="\'", end="\'")')


a = '5'
[辅助 repr；外层引号不是原输出] "a = '5'"


**逐项讲解：** 参数 `"a = "` 后有空格，sep 插入单引号，再输出 5，end 再加单引号，所以完整输出为 `a = '5'`，末尾没有换行。这两个单引号是输出内容；辅助 repr 外层的引号则只是描述记号。

> **English translation:**
> **Item explanation.** The argument `"a = "` ends with a space; `sep` inserts a single quote, then Python prints 5, and `end` appends another single quote. The complete output is `a = '5'` with no trailing newline. These two single quotes are output characters; the outer quotes shown by the helper's `repr` are only descriptive notation.


### 5.4

```python
a, b = 'b', 'a'
print(a, 'b', sep='>>>', end='\t')
print('a', b, sep='***', end='\n\n')
print(a, b, sep='\t')
```

In [39]:
a, b = 'b', 'a'
print(a, 'b', sep='>>>', end='\t')
print('a', b, sep='***', end='\n\n')
print(a, b, sep='\t')
show_invisible("a, b = 'b', 'a'\nprint(a, 'b', sep='>>>', end='\\t')\nprint('a', b, sep='***', end='\\n\\n')\nprint(a, b, sep='\\t')")


b>>>b	a***a

b	a

[辅助 repr；外层引号不是原输出] 'b>>>b\ta***a\n\nb\ta\n'


**逐项讲解：** 变量 a 保存 b，变量 b 保存 a。第 1 次输出 `b>>>b`，以 tab 结束；第 2 次紧接着输出 `a***a`，随后两个换行造成一空行；第 3 次输出 b、tab、a，再默认换行。完整 repr 是 `'b>>>b\ta***a\n\nb\ta\n'`。不要把变量名字当成其当前值，也不要假设每次 print 都单独占一行。

> **English translation:**
> **Item explanation.** Variable `a` stores `b`, while variable `b` stores `a`. The first call prints `b>>>b` and ends with a tab. The second continues immediately with `a***a`, then two newline characters create one blank line. The third prints `b`, a tab, and `a`, followed by its default newline. The complete `repr` is `'b>>>b\ta***a\n\nb\ta\n'`. Do not confuse variable names with their current values, and do not assume every `print` call occupies a separate line. Read `>>>` as “three greater-than signs,” `***` as “three asterisks,” `\t` as “backslash t,” and `\n` as “backslash n.”


## 6. 格式说明符：宽度、填充、对齐与精度

**原题（保留来源提取文本）**

```text
Exercise 6 | String formatting 
print(f"{12:8d}") 
 
print(f"{'12':0>8}") 
 
num = 12.34567890 
print(f"{num:8.3f}") 
 
print(f"{12.34567890:<08.3f}") 
 
name = 'dolphin' 
res1 = f'{name:->10}' 
res2 = f'{name:<10.3}' 
print(res1) 
print(res2)
```

**Knowledge references / 知识点出处**

- Lecture: [*Lecture 02*](../../lecture/Lecture%2002.pptx), Slide 48 (PPTX), “f-string” — **direct**；覆盖：f-string placeholders。
- Lecture: [*Lecture 02*](../../lecture/Lecture%2002.pptx), Slide 53 (PPTX), “f-string - number formatting” — **direct**；覆盖：numeric field width and precision。
- Lecture: [*Lecture 02*](../../lecture/Lecture%2002.pptx), Slide 51 (PPTX), “f-string – string formatting” — **direct**；覆盖：alignment / custom fill characters。
- Lecture: [*Lecture 02*](../../lecture/Lecture%2002.pptx), Slide 52 (PPTX), “f-string – string formatting” — **direct**；覆盖：string precision truncation。
- Lecture: [*Lecture 02*](../../lecture/Lecture%2002.pptx), Slide 53 (PPTX), “f-string - number formatting” — **prerequisite**；覆盖：zero padding with explicit alignment。 Shows zero-padded numeric fields without an explicit alignment character. Slide 51 supplies alignment background; the combination <08.3f is not directly demonstrated.
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 2, §2.10: Displaying Formatted Output with F-strings, printed p. 98 (PDF p. 99) — **direct**；覆盖：f-string placeholders。
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 2, §2.10: Displaying Formatted Output with F-strings, printed p. 100 (PDF p. 101) — **direct**；覆盖：numeric field width and precision。
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 2, §2.10: Displaying Formatted Output with F-strings, printed p. 102 (PDF p. 103) — **direct**；覆盖：numeric field width and precision。
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 2, §2.10: Displaying Formatted Output with F-strings, printed p. 105 (PDF p. 106) — **direct**；覆盖：alignment。
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 2, §2.10: Displaying Formatted Output with F-strings, printed p. 105 (PDF p. 106) — **prerequisite**；覆盖：custom fill characters / zero padding with explicit alignment。 Teaches alignment and width, but not arbitrary fill characters, zero padding, or zero padding combined with explicit alignment.
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 2, §2.10: Displaying Formatted Output with F-strings, printed p. 99 (PDF p. 100) — **prerequisite**；覆盖：string precision truncation。 Teaches the format-specifier grammar; this textbook section does not explain string precision as truncation.

原题来源：[recitation/Recitaion 02.pdf](../../recitation/Recitaion%2002.pdf#page=3)，PDF p. 3（物理页）。

### 本题考察什么

5 段程序分别检查：① 整数格式 d 与最小字段宽度；② 字符串配自定义填充及右对齐；③ 浮点数固定小数位和宽度；④ 显式左对齐与零填充的组合；⑤ 字符串右对齐填充，以及左对齐、截断和末尾空格。最后一段有两个输出，两个都要预测。

**怎么做：** 原题只有表格，讲义将任务解读为预测输出。先写值的类型，再依次标注填充、对齐、宽度、精度、格式类型；用竖线边界或 repr 检查不可见空格。

**来源与表述说明：** 自定义填充、字符串精度、零填充结合显式对齐的教材引用含 prerequisite；它们提供语法背景而不直接给出原题组合。教材 PDF p. 103 将 D 列为整数格式的说法有误，本题只使用有效的 d。

> **English translation:**
> ## 6. Format Specifiers: Width, Fill, Alignment, and Precision
>
> **Original exercise:** the English source text above is preserved verbatim.
>
> **What this exercise tests.** The five programs check: (1) integer format type `d` and minimum field width; (2) a string with a custom fill character and right alignment; (3) a float with fixed decimal places and field width; (4) explicit left alignment combined with zero filling; and (5) right-aligned string filling together with left alignment, truncation, and trailing spaces. The final program has two output lines, and both must be predicted.
>
> **How to proceed.** The source contains only a table, so the handout interprets the task as predicting output. First write the value's type, then identify fill, alignment, width, precision, and format type in that order. Use vertical-bar boundaries or `repr` to inspect invisible spaces. Read `|` as “vertical bar”; in a shell pipeline, the same symbol is usually called a “pipe.”
>
> **Source and wording note.** The textbook references for custom fill, string precision, and zero fill with explicit alignment are marked `prerequisite`: they provide syntax background without directly demonstrating the exact combinations in the exercise. Textbook PDF p. 103 incorrectly lists uppercase `D` as an integer format; this exercise uses the valid lowercase `d`. Read `:` as “colon,” `.` as “dot,” `<` as the left-align marker, and `>` as the right-align marker.


### 6.1

```python
print(f"{12:8d}")
```

In [40]:
print(f"{12:8d}")
show_invisible('print(f"{12:8d}")')


      12

[辅助 repr；外层引号不是原输出] '      12\n'


**逐项讲解：** `12` 是 int，`d` 用十进制整数显示；`8` 是最小总宽度，不是补 8 个字符。数字本身长 2，默认右对齐，所以前面补 6 个空格：`'      12'`。print 再追加换行。

> **English translation:**
> **Item explanation.** `12` is an `int`, and `d` displays it as a decimal integer. `8` is the minimum total width, not a request to add eight characters. The number itself has length 2 and is right-aligned by default, so six spaces are added before it: `'      12'`. `print` then appends a newline. Read `8d` as “width eight, decimal integer format.”


### 6.2

```python
print(f"{'12':0>8}")
```

In [41]:
print(f"{'12':0>8}")
show_invisible('print(f"{\'12\':0>8}")')


00000012

[辅助 repr；外层引号不是原输出] '00000012\n'


**逐项讲解：** `'12'` 是 str；`0` 是填充字符，`>` 是右对齐，`8` 是最小总宽度。左侧补 6 个 0，得到 `'00000012'`。这里的 0 是显示字符，不是改变字符串的数值类型。

> **English translation:**
> **Item explanation.** `'12'` is a `str`; `0` is the fill character, `>` is the right-align marker, and `8` is the minimum total width. Python adds six zeros on the left, producing `'00000012'`. The zeros are display characters; they do not change the string into a numeric type. Read `0>8` as “zero fill, right align, width eight.”


### 6.3

```python
num = 12.34567890
print(f"{num:8.3f}")
```

In [42]:
num = 12.34567890
print(f"{num:8.3f}")
show_invisible('num = 12.34567890\nprint(f"{num:8.3f}")')


  12.346

[辅助 repr；外层引号不是原输出] '  12.346\n'


**逐项讲解：** 先按 `.3f` 舍入到小数点后 3 位，得到 `12.346`（6 个字符，含小数点）；再按宽度 8 右对齐，前面补 2 个空格：`'  12.346'`。字段宽度包含整数、小数点与小数部分。

> **English translation:**
> **Item explanation.** First `.3f` rounds to three digits after the decimal point, giving `12.346` (six characters including the decimal point). Width 8 then right-aligns that text and adds two leading spaces: `'  12.346'`. Field width includes the integer part, decimal point, and fractional part. Read `.3f` as “dot three f” and `8.3f` as “width eight, precision three, fixed-point.”


### 6.4

```python
print(f"{12.34567890:<08.3f}")
```

In [43]:
print(f"{12.34567890:<08.3f}")
show_invisible('print(f"{12.34567890:<08.3f}")')


12.34600

[辅助 repr；外层引号不是原输出] '12.34600\n'


**逐项讲解：** `<` 明确左对齐，`0` 启用零填充，`8` 是宽度，`.3f` 仍然只负责舍入到 3 位小数。先得到 `12.346`，再在右边补 2 个 0 到宽度 8，结果为 `'12.34600'`。末尾的两个 0 是填充，不是多计算了两位小数；不能把结果解释为精度变成了 5。

> **English translation:**
> **Item explanation.** `<` explicitly selects left alignment, `0` enables zero filling, `8` is the width, and `.3f` still only rounds to three decimal places. Python first produces `12.346`, then adds two zeros on the right to reach width 8, giving `'12.34600'`. The final two zeros are fill, not two extra calculated decimal places; do not claim the precision became 5. Read `<08.3f` as “left align, zero fill, width eight, dot three f.”


### 6.5

```python
name = 'dolphin'
res1 = f'{name:->10}'
res2 = f'{name:<10.3}'
print(res1)
print(res2)
```

In [44]:
name = 'dolphin'
res1 = f'{name:->10}'
res2 = f'{name:<10.3}'
print(res1)
print(res2)
show_invisible("name = 'dolphin'\nres1 = f'{name:->10}'\nres2 = f'{name:<10.3}'\nprint(res1)\nprint(res2)")


---dolphin
dol       

[辅助 repr；外层引号不是原输出] '---dolphin\ndol       \n'


**逐项讲解：** 第一个格式 `->10`：用 `-` 填充、右对齐、宽度 10，dolphin 长 7，所以输出 `'---dolphin'`。第二个格式 `<10.3` 针对 str：`.3` 最多保留前三个字符 `dol`，再左对齐到宽度 10，右边补 7 个空格，输出 `'dol       '`。字符串 `.3` 是截断字符；浮点 `.3f` 是三位小数，浮点省略 f 的 `.3` 又是另一种有效数字规则。

> **English translation:**
> **Item explanation.** In the first specifier, `->10`, `-` is the fill character, `>` selects right alignment, and the width is 10. Because `dolphin` has length 7, the output is `'---dolphin'`. The second specifier, `<10.3`, applies to a string: `.3` keeps at most the first three characters, `dol`, and left alignment to width 10 adds seven spaces on the right, producing `'dol       '`. For a string, `.3` truncates characters; for a float, `.3f` means three decimal places; for a float without `f`, `.3` instead uses a significant-digit rule. Read `->10` as “dash fill, right align, width ten,” and `<10.3` as “left align, width ten, precision three.”


## 7. 销售利润预测：百分比是乘数

**原题（保留来源提取文本）**

```text
Exercise 7 | Sales Prediction  
A company has determined that its annual profit is typically 23 percent of total sales. 
Write a program that asks the user to enter the projected amount of total sales, and 
then displays the profit that will be made from that amount.
```

**Knowledge references / 知识点出处**

- Lecture: [*Lecture 02*](../../lecture/Lecture%2002.pptx), Slide 40 (PPTX), “Input function” — **direct**；覆盖：numeric keyboard input。
- Lecture: [*Lecture 02*](../../lecture/Lecture%2002.pptx), Slide 49 (PPTX), “f-string” — **direct**；覆盖：percentage calculation。 The example relates a fractional multiplier to a percentage; its printed f-string quotes and :2f explanatory label have typos.
- Lecture: [*Lecture 02*](../../lecture/Lecture%2002.pptx), Slide 44 (PPTX), “print function” — **direct**；覆盖：displaying computed values。
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 2, §2.6: Reading Input from the Keyboard, printed p. 78 (PDF p. 79) — **direct**；覆盖：numeric keyboard input。
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 2, §2.7: Performing Calculations, printed p. 82 (PDF p. 83) — **direct**；覆盖：percentage calculation / displaying computed values。

原题来源：[recitation/Recitaion 02.pdf](../../recitation/Recitaion%2002.pdf#page=4)，PDF p. 4（物理页）。

### 本题考察什么

这题要求把一句商业描述转成输入 → 处理 → 输出：读取可含小数的总销售额，把字符串转换成数值，用给定利润比例计算，并清楚显示金额。关键是百分数的含义，和“销售额”与“利润”两个量的区别。

**怎么做：** 使用可编辑的 `sales_text` 模拟 input 的字符串返回值，补全转换、计算和输出。完成后把该行改成注释所示 input，即为交互程序；暂不要求处理不合法输入。

> **English translation:**
> ## 7. Sales-Profit Prediction: A Percentage Is a Multiplier
>
> **Original exercise:** the English source text above is preserved verbatim.
>
> **What this exercise tests.** It asks you to translate a business description into input → processing → output: read total sales, which may contain decimals; convert the string to a number; calculate profit using the given rate; and display the amount clearly. The key distinction is between the meaning of a percentage and the separate quantities “sales” and “profit.”
>
> **How to proceed.** Use the editable `sales_text` value to simulate the string returned by `input`. Complete the conversion, calculation, and output. Then replace that line with the commented `input` version to obtain an interactive program. Handling invalid input is not required yet. Read `%` as “percent” in the problem statement, `/` as “divided by,” `$` as “dollar sign,” and `→` as “goes to.”


In [45]:
sales_text = "10000"  # 交互版：sales_text = input("预计总销售额：")
total_sales = float(sales_text)
profit = total_sales * 0.23
print(f"预计利润：${profit:.2f}")


预计利润：$2300.00


**推导与易错点：** 23% 表示每 100 单位销售额对应 23 单位利润，所以乘数是 `23 / 100 = 0.23`。输入先经过 `float` 才能进行数值计算。总销售额 10000 对应利润 2300，显示为 `$2300.00`。不要乘 23，也不要把总销售额乘 1.23；后者是把原数增加 23%，与题意不同。`.2f` 只控制金额显示，不应先把销售额转成 int 丢掉小数。

> **English translation:**
> **Derivation and common mistakes.** `23%` means 23 units of profit for every 100 units of sales, so the multiplier is `23 / 100 = 0.23`. The input must pass through `float` before numeric calculation. Total sales of 10,000 produce profit of 2,300, displayed as `$2300.00`. Do not multiply by 23, and do not multiply total sales by 1.23; the latter increases the original amount by 23%, which is a different meaning. `.2f` controls only the displayed currency precision; do not first convert sales to `int` and discard decimals. Read `%` as “percent,” `/` as “divided by,” `=` in the mathematical equation as “equals,” `$` as “dollar sign,” and `.2f` as “dot two f.”


## 8. 摄氏转华氏：翻译数学公式

**原题（保留来源提取文本）**

```text
Exercise 8 | Celsius to Fahrenheit Temperature Converter 
Write a program that converts Celsius temperatures to Fahrenheit temperatures. The 
formula is as follows: 
𝐹= 9
5 𝐶+ 32 
The program should ask the user to enter a temperature in Celsius, then display the 
temperature converted to Fahrenheit.
```

**Knowledge references / 知识点出处**

- Lecture: [*Lecture 02*](../../lecture/Lecture%2002.pptx), Slide 40 (PPTX), “Input function” — **direct**；覆盖：numeric keyboard input。
- Lecture: [*Lecture 02*](../../lecture/Lecture%2002.pptx), Slide 19 (PPTX), “Operators on int and float” — **prerequisite**；覆盖：translating a mathematical formula。 Teaches the arithmetic operators needed to implement the supplied formula, but does not teach Celsius conversion or the translation of mathematical notation.
- Lecture: [*Lecture 02*](../../lecture/Lecture%2002.pptx), Slide 44 (PPTX), “print function” — **direct**；覆盖：displaying computed values。
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 2, §2.6: Reading Input from the Keyboard, printed p. 78 (PDF p. 79) — **direct**；覆盖：numeric keyboard input。
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 2, §2.7: Performing Calculations, printed p. 88 (PDF p. 89) — **direct**；覆盖：translating a mathematical formula。
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 2, §2.7: Performing Calculations, printed p. 82 (PDF p. 83) — **direct**；覆盖：displaying computed values。

原题来源：[recitation/Recitaion 02.pdf](../../recitation/Recitaion%2002.pdf#page=4)，PDF p. 4（物理页）。

### 本题考察什么

这题检查输入转换、数学分式的程序表达、乘法符号必须显式写出，以及输出单位。你需要保留公式中的比例关系与偏移量，不能改变运算的组合。

**怎么做：** 先读下面恢复排版的公式，再用可编辑输入完成换算，并用容易心算的温度检验。

**来源与表述说明：** 原 PDF 将 9/5 排成上下分式，文本提取成为两行；上面的原题保留原始提取文本。下方仅恢复已核对的公式排版。

> **English translation:**
> ## 8. Celsius to Fahrenheit: Translating a Mathematical Formula
>
> **Original exercise:** the English source text above is preserved verbatim.
>
> **What this exercise tests.** It checks input conversion, translation of a mathematical fraction into Python, the need to write multiplication explicitly, and output units. Preserve both the ratio and offset in the formula; do not change how its operations are grouped.
>
> **How to proceed.** First read the correctly typeset formula below, then complete the conversion with the editable input value and test it using temperatures that are easy to calculate mentally.
>
> **Source and wording note.** The source PDF typesets 9/5 as a vertical fraction, which text extraction splits across two lines. The source block above preserves the extracted text; the next cell only restores the verified mathematical layout. Read `/` as “divided by,” `*` as “times,” `+` as “plus,” and `°` as “degrees.”


**恢复原分式的排版：**

$$F = \frac{9}{5}C + 32$$

其中 $C$ 是摄氏温度，$F$ 是华氏温度。这不是 `9 / (5*C + 32)`；写代码前先确认分数线的范围。

> **English translation:**
> **Restored layout of the original fraction:**
>
> Read the formula as “F equals nine fifths times C plus thirty-two.” Here $C$ is the Celsius temperature and $F$ is the Fahrenheit temperature. It is not `9 / (5*C + 32)`; confirm the scope of the fraction bar before writing code. Read `/` as “divided by,” `*` as “times,” `+` as “plus,” and the parentheses as “left/right parenthesis.”


In [46]:
celsius_text = "25"  # 交互版：celsius_text = input("摄氏温度：")
celsius = float(celsius_text)
fahrenheit = (9 / 5) * celsius + 32
print(f"{celsius:.1f} °C = {fahrenheit:.1f} °F")


25.0 °C = 77.0 °F


**推导与易错点：** 先把 `9/5` 作为比例，乘摄氏温度，再加 32；在 Python 中相邻的 `(9 / 5)` 和 `celsius` 之间必须写 `*`。`/` 会得到浮点比值 1.8，不能用 `//` 代替。示例 25 °C 对应 77 °F。可检查 0 °C → 32 °F、100 °C → 212 °F、−40 °C → −40 °F；这三个值同时帮助发现漏加 32、比例写反、误用整除等错误。

> **English translation:**
> **Derivation and common mistakes.** Treat `9/5` as a ratio, multiply it by the Celsius temperature, and then add 32. In Python, adjacent mathematical factors `(9 / 5)` and `celsius` must have an explicit `*` between them. `/` produces the floating-point ratio 1.8 and cannot be replaced with `//`. For example, 25 °C corresponds to 77 °F. Useful checks are 0 °C → 32 °F, 100 °C → 212 °F, and −40 °C → −40 °F; together they reveal a missing 32, a reversed ratio, or accidental floor division. Read `/` as “divided by,” `//` as “floor-divided by,” `*` as “times,” `°C` as “degrees Celsius,” `°F` as “degrees Fahrenheit,” and `→` as “becomes.”


## 9. 饼干配方调整：同一个比例用于三种原料

**原题（保留来源提取文本）**

```text
Exercise 9 | Ingredient Adjuster   
A cookie recipe calls for the following ingredients:  
1) 1.5 cups of sugar 
2) 1 cup of butter 
3) 2.75 cups of flour  
The recipe produces 48 cookies with this amount of the ingredients. 
Write a program that asks the user how many cookies he or she wants to make, and 
then displays the number of cups of each ingredient needed for the specified number of 
cookies.
```

**Knowledge references / 知识点出处**

- Lecture: [*Lecture 02*](../../lecture/Lecture%2002.pptx), Slide 40 (PPTX), “Input function” — **direct**；覆盖：numeric keyboard input。
- Lecture: [*Lecture 02*](../../lecture/Lecture%2002.pptx), Slide 36 (PPTX), “You Try it!” — **prerequisite**；覆盖：proportional ingredient scaling。 Uses price-times-quantity as proportional calculation background; it does not derive scaling from a 48-cookie reference recipe.
- Lecture: [*Lecture 02*](../../lecture/Lecture%2002.pptx), Slide 44 (PPTX), “print function” — **direct**；覆盖：displaying computed values。
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 2, §2.6: Reading Input from the Keyboard, printed p. 78 (PDF p. 79) — **direct**；覆盖：numeric keyboard input。
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 2, §2.7: Performing Calculations, printed p. 88 (PDF p. 89) — **prerequisite**；覆盖：proportional ingredient scaling。 Teaches multiplication, division, and formula translation, but does not derive the recipe scaling factor. The matching original textbook exercise is on PDF p.147, printed p.146.
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 2, §2.7: Performing Calculations, printed p. 82 (PDF p. 83) — **direct**；覆盖：displaying computed values。

原题来源：[recitation/Recitaion 02.pdf](../../recitation/Recitaion%2002.pdf#page=4)，PDF p. 4（物理页）。

### 本题考察什么

这题检查比例建模和变量复用。先从目标饼干数量与基准产量的关系得到缩放比例，再分别计算糖、黄油和面粉用量，并统一显示 cups 单位。题目里的 1)、2)、3) 是同一个配方的三种原料，不是三个独立应用。

**怎么做：** 先运行一次共同输入单元格，再按糖 → 黄油 → 面粉顺序完成三个计算单元格。改变目标数量后，从共同输入开始重跑本题；最终要得到一份含三种配料的完整结果。

**来源与表述说明：** 第 3 个编号在源块中带有配方产量与总体要求，这些共同要求已完整保留在原题中；三个编号共享同一目标数量与缩放比例。

> **English translation:**
> ## 9. Cookie Recipe Adjustment: One Ratio for Three Ingredients
>
> **Original exercise:** the English source text above is preserved verbatim.
>
> **What this exercise tests.** It checks proportional modeling and variable reuse. First derive a scale factor from the desired number of cookies and the base yield; then calculate the sugar, butter, and flour amounts separately and label all of them in cups. Items 1), 2), and 3) are three ingredients in one recipe, not three independent applications.
>
> **How to proceed.** Run the shared-input cell once, then complete the three calculation cells in the order sugar → butter → flour. After changing the desired quantity, rerun from the shared-input cell. The final result must describe one complete batch containing all three ingredients.
>
> **Source and wording note.** Source item 3 also contains the recipe yield and the overall instruction. Those shared requirements are fully preserved in the source block; all three numbered ingredients share one desired quantity and one scale factor. Read `/` as “divided by,” `*` as “times,” and `→` as “then” in the workflow.


**程序组织：** 先读目标数量，再确定缩放比例，最后依次输出糖、黄油、面粉。48 块是基准产量；目标数量应为非负整数，输入合法性检查留给后续课程。下面四个代码单元格合起来是一份顺序执行的程序。

> **English translation:**
> **Program organization.** Read the desired quantity, determine the scale factor, and then output sugar, butter, and flour in order. The base yield is 48 cookies. The desired quantity should be a nonnegative integer; input validation is left for a later course topic. Together, the four code cells below form one sequential program.


In [47]:
cookies_text = "24"  # 交互版：cookies_text = input("想制作多少块饼干：")
cookies = int(cookies_text)
scale = cookies / 48
print(f"制作 {cookies} 块饼干需要：")


制作 24 块饼干需要：


**共同推导：** 基准配方产出 48 块，所以目标配方是原配方的 `cookies / 48` 倍。目标 24 时比例为 0.5，目标 96 时比例为 2.0。比例必须使用 `/`，`//` 会把少于一批的正比例截成 0。输入数量用 int 合理，但配料量必须允许小数。

> **English translation:**
> **Shared derivation.** The base recipe makes 48 cookies, so the desired recipe is `cookies / 48` times the base recipe. The scale is 0.5 for 24 cookies and 2.0 for 96 cookies. The ratio must use `/` (read “divided by”); `//` (read “floor-divided by”) would reduce a positive fraction of one batch to 0. Using `int` for the cookie count is reasonable, but ingredient quantities must allow decimals.


### 9.1 糖（原题第 1 项）

使用共同缩放比例计算糖，并标明 cups。

> **English translation:**
> ### 9.1 Sugar (source item 1)
>
> Use the shared scale factor to calculate the sugar and label the result in cups.


In [48]:
sugar = 1.5 * scale
print(f"糖：{sugar:.3f} cups")


糖：0.750 cups


**本小项推导：** 糖的基准用量为 1.5 cups，因此乘共同的 scale。24 块时需要 0.750 cups；不要按增加/减少多少块去直接加减 1.5。

> **English translation:**
> **Subpart derivation.** The base amount of sugar is 1.5 cups, so multiply it by the shared `scale`. For 24 cookies, the result is 0.750 cups. Do not directly add or subtract 1.5 according to how many more or fewer cookies are requested. Read `*` as “times.”


### 9.2 黄油（原题第 2 项）

用同一目标数量与比例计算黄油。

> **English translation:**
> ### 9.2 Butter (source item 2)
>
> Use the same desired quantity and scale factor to calculate the butter.


In [49]:
butter = 1 * scale
print(f"黄油：{butter:.3f} cups")


黄油：0.500 cups


**本小项推导：** 黄油基准是 1 cup，因此用量为 `1 * scale`。24 块时是 0.500 cups。不要因为原始用量是整数就用 int 转换结果，那会丢失半杯。

> **English translation:**
> **Subpart derivation.** The base amount of butter is 1 cup, so the required amount is `1 * scale`. For 24 cookies, this is 0.500 cups. Do not convert the result with `int` merely because the original amount is a whole number; that would discard half a cup. Read `*` as “times.”


### 9.3 面粉（原题第 3 项）

完成最后一种原料，再检查三行输出是否对应同一批饼干。

> **English translation:**
> ### 9.3 Flour (source item 3)
>
> Complete the last ingredient, then check that all three output lines describe the same batch of cookies.


In [50]:
flour = 2.75 * scale
print(f"面粉：{flour:.3f} cups")


面粉：1.375 cups


**本小项推导：** 面粉的基准用量为 2.75 cups，同样乘 scale。24 块需要 1.375 cups。三种原料必须同时缩放；`.3f` 是为了清楚显示这个例子的杯数，不是题目规定的强制格式。改变输入后要从本题共同输入单元格重跑，不能让某种配料沿用旧比例。

> **English translation:**
> **Subpart derivation.** The base flour amount is 2.75 cups, so it is also multiplied by `scale`. For 24 cookies, the result is 1.375 cups. All three ingredients must scale together. `.3f` is used to display this example's cup amounts clearly; it is not a mandatory format imposed by the exercise. After changing the input, rerun from the shared-input cell so no ingredient uses an old scale factor. Read `.3f` as “dot three f.”


**教师：复制到 `.py` 后可直接交互的一份完整程序。** 此文本代码块不会在 Run All 时等待输入。

```python
cookies = int(input("想制作多少块饼干："))
scale = cookies / 48
sugar = 1.5 * scale
butter = 1 * scale
flour = 2.75 * scale
print(f"制作 {cookies} 块饼干需要：")
print(f"糖：{sugar:.3f} cups")
print(f"黄油：{butter:.3f} cups")
print(f"面粉：{flour:.3f} cups")
```

**代表值核对：** 24 块（半批）是 0.750 / 0.500 / 1.375 cups；48 块是 1.500 / 1.000 / 2.750 cups；96 块（双批）是 3.000 / 2.000 / 5.500 cups，顺序均为糖 / 黄油 / 面粉。0 块应全部为 0。不要只核对总和；每一种配料都应满足比例。

> **English translation:**
> **Instructor: a complete interactive program that can be copied into a `.py` file.** The code block above is text, so **Run All** does not wait for input.
>
> **Representative-value checks.** For 24 cookies (half a batch), sugar/butter/flour should be 0.750 / 0.500 / 1.375 cups. For 48 cookies, they should be 1.500 / 1.000 / 2.750 cups. For 96 cookies (two batches), they should be 3.000 / 2.000 / 5.500 cups. For 0 cookies, every amount should be 0. Do not check only the sum; each ingredient must satisfy the ratio. Read each `/` in the three-value list as “slash” (a visual separator), while `/` in `cookies / 48` is read “divided by.”


## 回顾与离堂检查

现在合上答案，用自己的话解释：

1. 为什么表达式的结果数字和 Python 类型都要检查？`/` 和 `//` 的差别是什么？
2. 遇到输出题时，你会按什么顺序检查类型、变量名、引号、`sep` 和 `end`？
3. 多重赋值的一行里，何时读取右侧旧值？两行之间又如何更新？
4. 字段宽度、数值精度与字符串精度分别控制什么？如何检查末尾空格？
5. 一个应用程序中，输入值、计算得到的值和显示格式有什么区别？你会选什么容易心算的输入检验公式？

**自查方法：** 能预测只是第一步；还要能说明每项为什么得到这个结果。若结果不符，先找第一步发生分歧的位置。完成 TODO 后重启内核、从头运行，确认程序不依赖上一次留下的变量。

> **English translation:**
> ## Recap and Exit Check
>
> Close the answers and explain in your own words:
>
> 1. Why must you check both the numeric result of an expression and its Python type? What is the difference between `/` (“slash,” true division) and `//` (“double slash,” floor division)?
> 2. For an output-prediction problem, in what order will you check types, variable names, quotation marks, `sep`, and `end`?
> 3. In one multiple-assignment statement, when are the old right-hand-side values read? How is the state updated between separate lines?
> 4. What do field width, numeric precision, and string precision control? How can you inspect trailing spaces?
> 5. In an application program, how do the input value, calculated value, and display format differ? Which mentally checkable inputs would you use to verify the formula?
>
> **Self-check method.** Prediction is only the first step; you must also explain why each result occurs. If the result differs, locate the first step where your reasoning and Python diverge. After completing the TODOs, restart the kernel and run from the beginning to confirm that the program does not depend on variables left from an earlier run.


**教师资料核对：原文件保持不变。**

- Lecture Slide 20：`10 // 0.25` 应为 `40.0`（浮点参与整除仍为 float）。
- Slide 24：`"Hello" + "ICP"` 应为 `HelloICP`，连接不自动插入字符。
- Slide 33：`z = 1` 后的跟随注释有误；Python 名字区分大小写，相关注释不可照抄。
- Slide 49：f-string 示例引号不匹配；固定两位小数需要 `.2f`，`:2f` 的 2 是宽度位置，不是两位小数精度。
- Textbook PDF p. 91 的概括要结合 p. 84 `/` 的类型例外；PDF p. 103 的大写 `D` 不是有效整数格式。本次只教实际使用的 `d`。

追问建议：让学生给出判断规则，再运行验证。E4 不要把“看见三个变量”误当成“保存着三个不同值”。E6 先解释各位置含义，再讨论组合规则；后续章节引用只供定位，不扩展为列表、文件或循环课。

> **English translation:**
> **Instructor source audit; the original files remain unchanged.**
>
> - Lecture Slide 20: `10 // 0.25` should produce `40.0` because floor division with a floating-point operand still returns a float. Read `//` as “floor-divided by.”
> - Slide 24: `"Hello" + "ICP"` should produce `HelloICP`; concatenation does not insert a character automatically. Read `+` here as “concatenated with.”
> - Slide 33: the comment following `z = 1` is incorrect. Python names are case-sensitive, so that comment should not be copied.
> - Slide 49: the f-string example has mismatched quotes. Exactly two decimal places require `.2f`; in `:2f`, the 2 occupies the width position, not the two-decimal precision position. Read `:` as “colon” and `.2f` as “dot two f.”
> - The general statement on textbook PDF p. 91 must be combined with the `/` type exception on p. 84. Uppercase `D` on PDF p. 103 is not a valid integer format; this recitation teaches the valid lowercase `d` only.
>
> Suggested follow-up questions: ask students to state a decision rule and then verify it by running code. In E4, do not let “three variables are visible” become “three distinct values are preserved.” In E6, explain the meaning of each format-specifier position before discussing combinations. Later-chapter citations are for locating concepts only; do not expand this recitation into lists, files, or loops.
